### DISCUSSION TITLE
Inc/Dec : hold for scrolling?

### DISCUSSION THREAD

**Discussion by subv0x** [RomanKubiak/ctrlr#557] [Q&A]
the following method does increment/decrement on the selected modulator eg; a combo list. 
what could be done so that holding down the click makes the inc/dec continuous? (so you don't have to click a million times)

```
incDecPCM = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
    local NAME=mod:getProperty("name")
    -- local endChar = string.sub(NAME,-1)
    local endChar = string.sub(NAME,-1)
    local begChar3 = string.sub(NAME,0,3)

    console(String("Name = : \""..NAME.."\""))
    console(String("endChar = : \""..endChar.."\""))
    console(String("begChar3 = : \""..begChar3.."\""))

    local comboName="pcm1_p"..endChar.."pt1"
    local val = panel:getModulatorByName(comboName):getValue()
    local min=panel:getModulatorByName(comboName):getProperty("modulatorMin")
    local max=panel:getModulatorByName(comboName):getProperty("modulatorMax")

    if begChar3 == "inc" then --  increment value
        if  val == max then --(this is the last element)
            ch=0
        else
            ch = val + 1
        end
    else -- decrement value
        if  val == min then --(this is the first element)
            ch=0
        else
            ch = val - 1
        end
    end
    panel:getModulatorByName(comboName):setValue(ch,true)
end
```

**Comment by damiensellier** [comment]
Hi,
Not sure if this is what you're looking for but when you select your uiButton that triggers the LUA code, go down to the bottom list of the component properties you'll find "keeping the button ON will produce a continuous stream of messages".
And next properties will set the interval between messages in this stream.

![PRESS AND HOLD BUTTON](https://github.com/RomanKubiak/ctrlr/assets/55164385/6f59761b-14f8-4bb4-98e3-dd8aa4522360)

You can also trigger events on different mouse behaviour (click, drag etc) on the component.
Maybe "mouse down" or "when mouse is click & dragged" could also fit your requirements:

![Mouse trigger behaviour](https://github.com/RomanKubiak/ctrlr/assets/55164385/6b9eb521-8d91-4ef2-bd5d-9419aaaf1f10)

**Reply by subv0x** [reply]
oh yes! very nice :tu: (supposed to be 'thumbs up') 185mS seems a good rate. note that it needs 'trigger button on mouse down' as well. 
i'm having some trouble with that method NOT working with it's 'twin' in a dual-view thing. 
wondering if it's local NAME=`mod:getProperty("name")` rather than component/owner - as discussed with dnaldoog yesterday. have to have another look at it. i've been copying some tabs-with-children and their methods into a fresh panel for a rebuild. going ok, but some unexpected hitches. 

the other thing is that the list loops going upwards eg: comes back round to 1, but doesn't cycle going down/decrementing. will have another go for the early morning coffee session ;-)

thanks !

**Reply by subv0x** [reply]
oh i remember why i haven't used that scrolling: it jumps values, hard to inc/dec by 1. 
... so that's going to mean getting to know more mousedown operations and writing a method for it. 
(going to check optimal rate, see if it can do it)
not forgetting this business about getting it to loop going downwards - makes more sense, if it loops going up.
waah so much to do ... real life calling, today. Ctrlr can be dangerous :-) if you have other stuff to attend to.

a click always increments by 2 steps minimum irrespective of speed. this is a Ctrlr problem. is Roman still maintaining it?
btw what is the optimal version to be using now? i have version:
Version = 5.3.201, Build date = Wed Mar 30 00:52:44 CEDT 2016, Branch = Nightly, Juce = 4.0.2, libusb = 1.0.19, liblo = 0.28. 
cannot remember what i was using in 2018.

**Reply by dnaldoog** [reply]
Probably best to post the panel as it's difficult to understand, for me at least, what you're trying to achieve!

5.3.201 is the version I use. There is a new version 5.6.3, I think it is, that also works. There is still no support for VST3 unless one jumps through hoops.

**Reply by subv0x** [reply]
ha yeh i'm going to start a thread for it, and bust everyone's brains with it again :D 
'D110 Again'.

i don't really care too much about VST3. i have a VST2 Steinberg licence, from when they stopped it, because i was a synthedit user. it really messed everything up, 64bit and VST3 coming at the same time. VST3 isn't really relevant for synth editors ????!!!! lol nor is 64bit. 
Synthedit v1.017 rocks ! never got into the later versions, it was all the wrong way round for the logic and GUI stuff.
